### PR TITLE
[SourceKit] In related identes, report underlying var of closure capture and shorthand if let binding

### PR DESCRIPTION
**Pull request by ahoppen**
Previously, the related idents request wouldn’t look through caputred variables like `[foo]`. Change the logic to consider the captured variable as well as the variable that’s implicitly declared for use inside the closure. Do the same logic for shorthand `if let foo {` bindings.

rdar://81628899

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 1]
```diff
@@ -2190,6 +2195,81 @@ class RelatedIdScanner : public SourceEntityWalker {
...
+          continue;
+        }
+
+        assert(DeclaredVar->getName() == ReferencedVar->getName());
+        if (DeclaredVar == Dcl) {
+          RelatedDecls.push_back(ReferencedVar);
+        }
+        if (ReferencedVar == Dcl) {
+          RelatedDecls.push_back(DeclaredVar);
+        }
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2232]: This should be an either or right?

**Review by bnbarham — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 2]
```diff
@@ -2301,21 +2380,54 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
...
+        // reference of the captured variable and the declaration of the
+        // variable usable in the closure.
+        llvm::SmallDenseSet<std::pair<unsigned, unsigned>, 8> RangesSet;
+
+        // List of decls whose ranges should be reported as related identifiers.
+        SmallVector<ValueDecl *, 2> Worklist;
+        Worklist.push_back(VD);
+
+        // Decls that we have already visited, so we don't walk circles.
+        SmallPtrSet<ValueDecl *, 2> VisitedDecls;
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2394]: LLVM has `SmallSetVector` which you could use instead

Comment (ahoppen) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2394]: Why do you think `SmallSetVector` would be better than a `SmallPtrSet` here? I don’t think we need any ordering guarantees here.

Comment (bnbarham) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2394]: The idea would be to merge `Worklist` and `VisitedDecls`.

Comment (bnbarham) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2394]: Ah, sorry, SetVector also removes from the `Set` in a `pop_back`. Ignore me.

**Pull request merged**